File: Android/SheetMusicAudioAndroid/src/main/kotlin/io/github/jiyimeta/sheetmusic/audio/synth/MasterTuning.kt
package io.github.jiyimeta.sheetmusic.audio.synth

import kotlin.math.roundToInt

/**
 * A4 master-tuning math, mirroring Swift `MasterTuning` in SheetMusicAudioCore.
 * FluidSynth honors MIDI Master Tuning RPNs, so Android retunes by sending these
 * CC pairs to each melodic channel. Kept in lockstep with the Swift helper by
 * [MasterTuningTest]'s golden assertions.
 */
internal object MasterTuning {
    data class CC(val controller: Int, val value: Int)

    /** Split a cents offset (from A4=440) into nearest-semitone coarse + remaining fine cents. */
    fun split(cents: Double): Pair<Int, Double> {
        val coarse = (cents / 100.0).roundToInt()
        return coarse to (cents - 100.0 * coarse)
    }

    /** CC pairs (in order) to retune one channel by [cents] on an RPN-honoring synth. */
    fun rpnControlChanges(cents: Double): List<CC> {
        val (coarse, fineCents) = split(cents)
        val fine14 = (8192 + (fineCents / 100.0 * 8192.0).roundToInt()).coerceIn(0, 16383)
        val coarseValue = (64 + coarse).coerceIn(0, 127)
        return listOf(
            CC(101, 0), CC(100, 2), CC(6, coarseValue), CC(38, 0),
            CC(101, 0), CC(100, 1), CC(6, fine14 ushr 7), CC(38, fine14 and 0x7F),
            CC(101, 127), CC(100, 127),
        )
    }
}

File: Sources/SheetMusicAudioCore/MasterTuning.swift
import Foundation

/// A4 master-tuning math shared by the Apple and Android playback engines.
///
/// A cents offset from A4=440 is split into a whole-semitone "coarse" part and a
/// remaining "fine" part. iOS feeds the split straight into the AUMIDISynth's
/// global AudioUnit tuning params (id 901 Coarse / 902 Fine); Android turns it
/// into the MIDI Master Tuning RPN that FluidSynth honors. Keeping the split here
/// gives both platforms one source of truth.
public enum MasterTuning { // swiftlint:disable:this inclusive_language
    /// Split a cents offset (relative to A4=440) into nearest-semitone coarse +
    /// remaining fine cents. `coarse*100 + fine == cents`, with `|fine| ≤ 50`.
    public static func split(cents: Double) -> (coarseSemitones: Int, fineCents: Double) {
        let coarse = Int((cents / 100).rounded())
        return (coarse, cents - 100 * Double(coarse))
    }

    /// One MIDI Control-Change message (controller + 7-bit value).
    public struct CC: Equatable, Sendable {
        public let controller: UInt8
        public let value: UInt8
        public init(controller: UInt8, value: UInt8) {
            self.controller = controller
            self.value = value
        }
    }

    /// CC pairs to send (in order) to one channel so an RPN-honoring synth
    /// (e.g. FluidSynth) retunes by `cents`. Both data-entry bytes are emitted
    /// per RPN so 14-bit-strict synths apply the value.
    public static func rpnControlChanges(cents: Double) -> [CC] {
        let (coarse, fineCents) = split(cents: cents)
        let fine14 = max(0, min(16383, 8192 + Int((fineCents / 100 * 8192).rounded())))
        let coarseValue = UInt8(max(0, min(127, 64 + coarse)))
        return [
            CC(controller: 101, value: 0), CC(controller: 100, value: 2),
            CC(controller: 6, value: coarseValue), CC(controller: 38, value: 0),
            CC(controller: 101, value: 0), CC(controller: 100, value: 1),
            CC(controller: 6, value: UInt8(fine14 >> 7)), CC(controller: 38, value: UInt8(fine14 & 0x7F)),
            CC(controller: 101, value: 127), CC(controller: 100, value: 127),
        ]
    }
}
